### PR TITLE
refactor: [M3-7377] - Migrate dopdowns in Create Firewall Drawer from Autocomplete to LinodeSelect and NodeBalancerSelect

### PR DESCRIPTION
**Pull request by tyler-akamai**
## Description 📝
Migrated dopdowns in Create Firewall Drawer from Autocomplete to LinodeSelect and NodeBalancerSelect.

## Changes  🔄
List any change relevant to the reviewer.
- Migrated dopdowns in Create Firewall Drawer from Autocomplete to LinodeSelect and NodeBalancerSelect.
- Updated LinodeSelect to use new autocomplete component: `src/components/Autocomplete/Autocomplete`
- Updated NodeSelect to use new autocomplete component
- Added unit tests (based on `LinodeSelect.test.tsx`) for `NodeBalancerSelect.tsx`

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| Screenshot 2023-11-09 at 11 23 55 AM | Screenshot 2023-11-15 at 2 27 10 PM |
| Screenshot 2023-11-09 at 11 24 22 AM | Screenshot 2023-11-15 at 2 27 56 PM |

### Prerequisites
(How to setup test environment)
- There are no prerequisites

### Reproduction steps
(How to reproduce the issue, if applicable)
- Since this is a refactor, there are no reproduction steps

### Verification steps 
(How to verify changes)

- [ ] Navigate to http://localhost:3000/linodes/create and ensure when clicking on 'Create Firewall' in the Firewall Panel, the label for the Linodes dropdown is 'Additional Linodes (Optional)'
- [ ] Navigate to http://localhost:3000/nodebalancers/create and ensure when clicking on 'Create Firewall' in the Firewall Panel, the label for the NodeBalancers dropdown is 'Additional NodeBalancers (Optional)'

Ensure the LinodeSelect and NodeBalancerSelect dropdowns are working:
- [ ] http://localhost:3000/firewalls/create
- [ ] http://localhost:3000/domains/create

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [x] 📑 Providing or updating our documentation
- [x] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [x] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx [hunk 1]
```diff
@@ -1,88 +1,178 @@
 import { NodeBalancer } from '@linode/api-v4';
-import Autocomplete from '@mui/material/Autocomplete';
+import { APIError } from '@linode/api-v4/lib/types';
```
Comment (tyler-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx:2]: I used the LinodeSelect file as a reference when updating this file to use the new Autocomplete component.

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx [hunk 2]
```diff
@@ -40,7 +39,7 @@ export const READ_ONLY_DEVICES_HIDDEN_MESSAGE =
   'Only services you have permission to modify are shown.';
 
 export interface CreateFirewallDrawerProps {
-  inCreateFlow?: boolean;
+  createFlow?: FirewallDeviceEntityType;
```
Comment (tyler-akamai) [inline packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx:42]: `undefined` - Opened from Firewall Landing Page
`linode` - Opened from Create Linode Page
`nodebalancer - Opened from Create NodeBalancer Page

if `undefined`, the labels will be 'Linodes' and 'NodeBalancers'
if `linode`, the label will be 'Additional Linodes (Optional)'
if `nodebalancer`, the label will be 'Additional NodeBalancers (Optional)'

File: packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx [hunk 3]
```diff
@@ -1,88 +1,178 @@
...
+    sx,
+    value,
+  } = props;
 
-  const options = nodebalancers?.map(({ id, label }) => ({ id, label }));
+  const [inputValue, setInputValue] = React.useState('');
 
-  const selectedNodebalancer =
-    options?.find((option) => option.id === value) ?? null;
+  const { data, error, isLoading } = useAllNodeBalancersQuery();
```
Comment (tyler-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx:97]: This query has a comment above it that says 'do not use'

it seems to be working fine, does anyone have any thoughts on this?

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 4]
```diff
@@ -16,12 +16,16 @@ interface Props {
   handleFirewallChange: (firewallID: number) => void;
   helperText: JSX.Element;
   selectedFirewallId: number;
+  serviceType?: FirewallDeviceEntityType;
```
Comment (hana-linode) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:19]: Can we rename this to `entityType` instead to remain consistent?

File: packages/manager/src/features/NodeBalancers/NodeBalancerSelect.test.tsx [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    // Wait for the options to load (use some unique identifier for the options)
+    await waitFor(() => {
+      const customOption = screen.getByTestId('custom-option-1');
+      expect(customOption).toBeInTheDocument();
+      expect(customOption).toHaveTextContent(
+        'metadata-test-region - Not Selected'
+      );
+    });
+  });
+  test('should display custom no options message', async () => {
```
Comment (hana-linode) [inline packages/manager/src/features/NodeBalancers/NodeBalancerSelect.test.tsx:56]: I think using `it` over `test` reads better
```suggestion
  it('should display custom no options message if one is passed in', async () => {
```

File: packages/manager/src/features/NodeBalancers/NodeBalancerSelect.test.tsx [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    // Open the dropdown
+    userEvent.click(input);
+
+    await waitFor(() => {
+      // The custom no options message should be displayed when there are no options available
+      expect(screen.getByText(customNoOptionsMessage)).toBeInTheDocument();
+    });
+  });
+
+  test('should display default no options message', async () => {
```
Comment (hana-linode) [inline packages/manager/src/features/NodeBalancers/NodeBalancerSelect.test.tsx:82]:
```suggestion
  it('should display default no options message if no custom message is passed', async () => {
```

File: packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx (hunk 3 above)
Comment (bnussman-akamai) [reply packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx:98]: We say `Please don't use` to try to prevent instances where we fetch all entities. The alternative to this would be to build an infinitly-scollable select like `VolumeSelect`. It would be nice to have, but there are pros and cons to both approaches. Using a fetch-all here is probably find, but one day we might explore making infinitly-scollable selects more standard.

**Review by hana-linode — COMMENTED** [review]
The Optional text for the Firewall drawer is a bit misleading since I was able to create a Firewall without a Linode/Nodebalancer. It seems like both options are optional?

image

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx
Comment (bnussman-akamai) [inline packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx:1]: This might be out of scope of this ticket but the UX of saying `Linodes (Optional)` and `NodeBalancers` (without optional) feels weird to me. It makes it unclear if NodeBalancers is required or optional.

Screenshot 2023-11-15 at 9 28 28 AM

**Review by bnussman-akamai — COMMENTED** [review]
Do we know if the e2es are passing? We may need to push to kick-off a job

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx
Comment (tyler-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.tsx:1]: Yes this is a good catch.

Talked to UI:
"Our general rule is to call out the outlier(s). For example, in a given form, if most inputs are required add (optional) to label of the one(s) that are optional.  If most inputs are optional add (required) to the label of the one(s) that are required.  In your example above, the "Label" input should be called out as required.   Btw., the font weight for the optional/required text should be regular."

so basically I changed it to:
Screenshot 2023-11-15 at 2 24 44 PM
Screenshot 2023-11-15 at 2 25 19 PM
Screenshot 2023-11-15 at 2 25 37 PM

**Review by carrillo-erik — APPROVED** [review]

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.test.tsx [hunk 7]
```diff
@@ -28,26 +35,45 @@ describe('Create Firewall Drawer', () => {
...
     act(() => {
-      userEvent.type(screen.getByLabelText('Label'), 'test label');
-      userEvent.type(screen.getByLabelText('Linodes'), 'test linode');
-      userEvent.type(
-        screen.getByLabelText('NodeBalancers'),
-        'test nodebalancer'
-      );
+      userEvent.type(screen.getByLabelText('Label (required)'), options.label);
+      userEvent.type(screen.getByLabelText('Linodes'), 'linode1');
+      // userEvent.type(screen.getByLabelText('NodeBalancers'), 'nodebalancer1');
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.test.tsx:76]: I don't think these lines do anything meaningful. To select an item in an Autocomplete, you need to click on an item or press `enter` on it. I don't think this is doing what it appears to be doing. The test probably passes both Linodes and NodeBalancers are optional. 

I'm thinking we should either find a way to select an Autocomplete item or just handle this with an e2e test later on.

File: packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddLinodeDrawer.tsx [hunk 8]
```diff
@@ -66,7 +69,7 @@ export const AddLinodeDrawer = (props: Props) => {
       const label = selectedLinodes[index].label;
       const id = selectedLinodes[index].id;
       if (result.status === 'fulfilled') {
-        enqueueSnackbar(`Linode ${label} added successfully`, {
+        enqueueSnackbar(`Linode ${label} successfully added`, {
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddLinodeDrawer.tsx:72]: Does this need a `.`?  I think this is considered a sentence

File: packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx [hunk 9]
```diff
@@ -1,88 +1,185 @@
...
+            }
+          : undefined
+      }
+      value={
+        typeof value === 'function'
+          ? multiple && Array.isArray(value)
+            ? nodebalancers?.filter(value) ?? null
+            : nodebalancers?.find(value) ?? null
+          : mapIdsToDevices<NodeBalancer>(value, nodebalancers)
+      }
```
Comment (bnussman-akamai) [inline packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx:149]: It might be overkill to have this component handle value being of type `function`. The Linode Select needed that for a special case. 

Doesn't hurt to leave it though.

**Review by bnussman-akamai — APPROVED** [review]
Functionality looks good!

**Review by hana-linode — APPROVED** [review]
✅ Navigate to http://localhost:3000/linodes/create and ensure when clicking on 'Create Firewall' in the Firewall Panel, the label for the Linodes dropdown is 'Additional Linodes (Optional)'
✅ Navigate to http://localhost:3000/nodebalancers/create and ensure when clicking on 'Create Firewall' in the Firewall Panel, the label for the NodeBalancers dropdown is 'Additional NodeBalancers (Optional)'

Ensure the LinodeSelect and NodeBalancerSelect dropdowns are working:
✅ http://localhost:3000/firewalls/create
✅ http://localhost:3000/domains/create

**Review by jdamore-linode — APPROVED** [review]

File: packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx (hunk 9 above)
Comment (tyler-akamai) [reply packages/manager/src/features/NodeBalancers/NodeBalancerSelect.tsx:149]: Yeah it seems like its for the `ConfigNodeIPSelect` file. I am going to leave it for now, but I can circle back around to it.

File: packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.test.tsx (hunk 7 above)
Comment (tyler-akamai) [reply packages/manager/src/features/Firewalls/FirewallLanding/CreateFirewallDrawer.test.tsx:76]: Sounds good, we are going to be writing e2e tests soon, so I will make a note to add this. Good catch!
